experimental/tools/handshake-simulator/ExecModels.cpp

```diff
@@ -68,59 +96,50 @@ static SmallVector<Value> toVector(ValueRange range) {
 
 /// Returns whether the precondition holds for a general op to execute
 static bool isReadyToExecute(ArrayRef<Value> ins, ArrayRef<Value> outs,
-                             llvm::DenseMap<Value, llvm::Any> &valueMap) {
+                             ChannelMap &channelMap) {
+
   for (auto in : ins)
-    if (valueMap.count(in) == 0)
+    if (channelMap[in].state == DataflowState::NONE)
       return false;
 
   for (auto out : outs)
-    if (valueMap.count(out) > 0)
+    if (channelMap[out].state == DataflowState::VALID)
       return false;
 
   return true;
 }
 
 /// Fetches values from the value map and consume them
-static std::vector<llvm::Any>
-fetchValues(ArrayRef<Value> values,
-            llvm::DenseMap<Value, llvm::Any> &valueMap) {
+static std::vector<llvm::Any> fetchValues(ArrayRef<Value> values,
+                                          ChannelMap &channelMap) {
   std::vector<llvm::Any> ins;
   for (auto &value : values) {
-    assert(valueMap[value].has_value());
-    ins.push_back(valueMap[value]);
-    // valueMap.erase(value);
+    if (channelMap[value].state == DataflowState::VALID)
+      ins.push_back(channelMap[value].data.value());
```

Comment:
nit: whenever you know that an optional has a value, you can just do `*(channelMap[value].data)` (I think the outer parenthesis are necessary here, check if that's true) instead of `channelMap[value].data.value()`, which is slightly shorter